poc_mobile_verifier/src/verifier.rs

```diff
@@ -0,0 +1,176 @@
+            // submit rewards
+            if epoch_since_last_reward_duration >= reward_period {
+                tracing::info!("Rewarding epoch: {:?}", epoch_since_last_reward);
+                self.reward_epoch(epoch_since_last_reward).await?
+            } else if epoch_since_last_reward_duration + sleep_duration >= reward_period {
+                // If the next epoch is a reward period, cut off sleep duration.
+                // This ensures that verifying will always end up being aligned with
+                // the desired reward period.
+                sleep_duration = reward_period - epoch_since_last_reward_duration;
+            }
+
+            tracing::info!("Sleeping...");
+            let shutdown = shutdown.clone();
+            tokio::select! {
+                _ = shutdown => return Ok(()),
+                _ = sleep(
+                    sleep_duration
+                        .to_std()
+                        .map_err(|_| Error::OutOfRangeError)?,
+                ) => (),
+            }
+        }
+    }
+
+    pub async fn verify_epoch(&mut self, epoch: Range<DateTime<Utc>>) -> Result {
+        let shares = self.verifier.verify_epoch(&epoch).await?;
+
+        let mut transaction = self.pool.begin().await?;
+
+        // Should we remove the heartbeats that were not new
+        // from valid shares
+        // TODO: switch to a bulk transaction
+        for share in shares.valid_shares.clone() {
+            let heartbeat = Heartbeat::from(share);
+            heartbeat.save(&mut transaction).await?;
+        }
+
+        // Update the last verified end time:
+        self.last_verified_end_time
+            .update(&mut transaction, epoch.end.timestamp() as i64)
+            .await?;
+
+        transaction.commit().await?;
+
+        shares
+            .write(&self.valid_shares_tx, &self.invalid_shares_tx)
+            .await?;
+
+        Ok(())
+    }
+
+    pub async fn reward_epoch(&mut self, epoch: Range<DateTime<Utc>>) -> Result {
+        let heartbeats = Heartbeats::validated(&self.pool, epoch.start).await?;
+
+        let rewards = self.verifier.reward_epoch(&epoch, heartbeats).await?;
+
+        let mut transaction = self.pool.begin().await?;
+
+        // Clear the heartbeats database
+        // TODO: should the truncation be bound to a given epoch?
+        // It's not intended that any heartbeats will exists outside the
+        // current epoch, but it might be better to code defensively.
+        sqlx::query("TRUNCATE TABLE heartbeats;")
+            .execute(&mut transaction)
+            .await?;
+
+        // Update the last rewarded end time:
+        self.last_rewarded_end_time
+            .update(&mut transaction, epoch.end.timestamp() as i64)
+            .await?;
+
+        transaction.commit().await?;
+
+        rewards
+            .write(&self.subnet_rewards_tx)
+            .await?
+            // Await the returned one shot to ensure that we wrote the file
+            .await??;
```

Comment:
just verifying this isn't a merge confict resolution gone wrong?

Reply:
Yup, this is correct. One result for awaiting the oneshot, one result for if the file was written properly.

Reply:
What about the first `await?`

Reply:
That is for sending the bytes over the channel to the file sink